src/cdi_ndi/NDI_Base.hh

```diff
@@ -0,0 +1,166 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
+ * \file   cdi_ndi/NDI_Base.hh
+ * \author Ben R. Ryan
+ * \date   2019 Nov 4
+ * \brief  NDI_Base class definition.
+ * \note   Copyright (C) 2019-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//----------------------------------------------------------------------------//
+
+#ifndef cdi_ndi_NDI_Base_hh
+#define cdi_ndi_NDI_Base_hh
+
+#include "NDI_Common.hh"
+#include "ndi.h"
+#include "ds++/Assert.hh"
+#include <algorithm>
+#include <map>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace rtt_cdi_ndi {
+
+//============================================================================//
+/*!
+ * \class NDI_Base
+ *
+ * \brief Base class for wrapping NDI calls to NDI data. Reads data, constructs
+ *        internal storage amenable to radiation calculations, and provides
+ *        accessors. Instantiated only through a data type-specific derived
+ *        class.
+ */
+//============================================================================//
+
+class NDI_Base {
+
+protected:
+  //! Path to gendir file, which indexes an NDI dataset
+  const std::string gendir;
+  //! Type of data to read (NDI supports multigroup_neutron, multigroup_photon,
+  //! multigroup_multi, tn, tnreactions, and dosimetry_neutrons)
+  const std::string dataset;
+  //! Name of library in which to find reaction
+  const std::string library;
+  //! Name of reaction to read
+  const std::string reaction;
+  //! Multigroup or continuous energy data
+  const DISCRETIZATION discretization;
+
+  //! Constructor
+  NDI_Base(const std::string &gendir_in, const std::string &dataset_in,
+           const std::string &library_in, const std::string &reaction_in,
+           const DISCRETIZATION discretization_in);
+
+  //! Name of reaction as found in NDI data
+  std::string reaction_name;
+
+  //! Labels (ZAIDs) for reaction products
+  std::vector<int> products;
+
+  //! Map from reaction product ZAID to index
+  std::map<int, int> product_zaid_to_index;
+
+  //! Multiplicities for each reaction product
+  std::vector<int> product_multiplicities;
+
+  //! Temperature support point grid for reaction
+  std::vector<double> reaction_temperature;
+
+  //! Incident energy support point grid for reaction
+  std::vector<double> einbar;
+
+  //! Incident cross section support point grid for reaction
+  std::vector<double> sigvbar;
+
+  //! Temperature support point grids for each reaction product
+  std::vector<std::vector<double>> product_temperatures;
+
+  //! Distribution support point grids for each reaction product
+  std::vector<std::vector<std::vector<double>>> product_distributions;
```

Comment:
In general this array can be ragged-right. This is definitely the case for continuous energy, but I think it can also be true for multigroup (different support point spacing in material temperature).